Polyfills/Canvas/Source/nanovg/nanovg_filterstack.cpp

```diff
@@ -141,36 +193,110 @@ void nanovg_filterstack::Render(
-                setUniform(m_uniforms.u_direction, horizontal);
+                if (element.blurElement.horizontal < 2)
+                {
+                    std::vector<float> kernel = CalculateGaussianKernel(element.blurElement.horizontal, BLUR_TAPS);
+                    float horizontal[4] = {1.f, 0.f, 0.f, 0.f};
+                    setUniform(m_uniforms.u_direction, horizontal, 1);
+                    setUniform(m_uniforms.u_weights, kernel.data(), BLUR_UNIFORM_SIZE);
 
-                nextBuf = acquire();
-                filterPass(blurProg, prevBuf, nextBuf);
-                release(prevBuf);
-                prevBuf = nextBuf;
-                nextBuf = nullptr;
+                    nextBuf = acquire();
+                    filterPass(gaussBlurProg, prevBuf, nextBuf);
+                    release(prevBuf);
+                    prevBuf = nextBuf;
+                    nextBuf = nullptr;
+                }
+                else
+                {
+                    std::array<float, 2> kernel = CalculateBoxKernel(element.blurElement.horizontal);
+                    bool isOdd = static_cast<int>(kernel[0]) % 2 == 1;
+                    std::array<std::array<float, 4>, 3> kernelsOdd = {
+                        std::array<float, 4>{kernel[0], kernel[1], 0.f, 0.f},
+                        std::array<float, 4>{kernel[0], kernel[1], 0.f, 0.f},
+                        std::array<float, 4>{kernel[0], kernel[1], 0.f, 0.f},
+                    };
+                    std::array<std::array<float, 4>, 3> kernelsEven = {
+                        std::array<float, 4>{kernel[0], kernel[1], -0.5f, 0.f},
+                        std::array<float, 4>{kernel[0], kernel[1], 0.5f, 0.f},
+                        std::array<float, 4>{kernel[0] + 1, kernel[1], 0.f, 0.f},
+                    };
+                    float horizontal[4] = {1.f, 0.f, 0.f, 0.f};
+
+                    // 3 pass box blur for s >= 2
+                    for (int i = 0; i < 3; i++)
+                    {
+                        setUniform(m_uniforms.u_direction, horizontal, 1);
+                        setUniform(m_uniforms.u_weights, isOdd ? kernelsOdd[i].data() : kernelsEven[i].data(), 1);
+
+                        nextBuf = acquire();
+                        filterPass(boxBlurProg, prevBuf, nextBuf);
+                        release(prevBuf);
+                        prevBuf = nextBuf;
+                        nextBuf = nullptr;
+                    }
+                }
 
                 // Vertical Pass
-                float vertical[4] = {0.f, element.blurElement.vertical * 1.f, 0.f, 0.f}; // scale by element.blurElement.vertical
-                setUniform(m_uniforms.u_direction, vertical);
-                if (last)
+                if (element.blurElement.vertical < 2)
+                {
+                    std::vector<float> kernel = CalculateGaussianKernel(element.blurElement.vertical, BLUR_TAPS);
```

Comment:
same as previous code block. Is it possible to use 2 directions array `float directions[2][4]= {{1.f,0.,0.,0.},[0.,1.,0.,0.}} ` and replace with a loop ?